src/main/java/org/miracum/etl/fhirtoomop/repository/service/PatientMapperServiceImpl.java

```diff
@@ -45,6 +66,74 @@ public Long findPersonIdByFhirLogicalId(String logicalId) {
     return existingPerson.getPersonId();
   }
 
+  /**
+   * Queries HAPI FHIR MDM for the golden resource ID
+   *
+   * @param logicalId logical id of the Patient resource
+   * @return
+   */
+  public String getGoldenResourceByFhirLogicalId(String logicalId) {
+    return logicalIdToGolden.get(
+        logicalId,
+        key -> {
+          String fhirId = StringUtils.prependIfMissing(key.replaceFirst("pat-", ""), "Patient/");
+
+          // Query for links with this id
+          Parameters response =
+              client
+                  .operation()
+                  .onServer()
+                  .named("$mdm-query-links")
+                  .withParameter(Parameters.class, "resourceId", new StringType(fhirId))
+                  .execute();
+
+          // Get the link param lists
+          List<List<Parameters.ParametersParameterComponent>> links =
+              response.getParameter().stream()
+                  .filter(param -> param.getName().equals("link"))
+                  .map(
+                      linkParams -> {
+                        return linkParams.getPart();
+                      })
+                  .toList();
+
+          // Find the link with MATCH or POSSIBLE_MATCH
+          for (List<Parameters.ParametersParameterComponent> link : links) {
+            Optional<String> goldenResourceId = getGoldenResourceFromLink(link);
+            if (goldenResourceId.isPresent()) {
+              return "pat-" + goldenResourceId.get().replaceFirst("Patient/", "");
+            }
+          }
+
+          // If we did not find a MATCH then we return the same ID
+          return key;
+        });
+  }
+
+  private Optional<String> getGoldenResourceFromLink(
+      List<Parameters.ParametersParameterComponent> link) {
+    for (Parameters.ParametersParameterComponent param : link) {
+      if (param.getName().equals("matchResult")) {
+        String value = param.getValue().toString();
+        if (value.equals("MATCH")
+            || (treatPossibleMatchesAsMatch && value.equals("POSSIBLE_MATCH"))) {
```

Comment:
Is this method now only called when `treatPossibleMatchesAsMatch` is `false`, so won't this method just be a no-op for resources without a MATCH golden?

Reply:
No, since non golden resources still have to use it to update their references.

Reply:
You have to do this to check if it's a POSSIBLE_MATCH, though?